# Project: Automated Feature Engineering Workflow

This project demonstrates automated feature engineering for a simple regression problem.
It uses a preprocessing pipeline with:

- OneHotEncoder for categorical variables
- Passthrough for numeric features
- RandomForest for prediction

## Steps:
1. Load dataset
2. Split into training/testing sets
3. Automatically transform numerical + categorical features
4. Train RandomForest regression model
5. Evaluate using RMSE

## Files:
- automated_feature_engineering.ipynb
- sales.csv
- requirements.txt

In [22]:
# Import packages and libraries

import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.metrics import mean_squared_error
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error


In [23]:
# Load dataset
df = pd.read_csv("sales.csv")
df.head()

Region  Customers  Marketing_Spend  Sales
0  North        120             5000  15000
1  South         80             3000   9000
2   East         95             4000  11000
3   West        110             5200  14500
4  North        130             5500  16000

In [24]:
df.shape

(152, 4)

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 152 entries, 0 to 151
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Region           152 non-null    object
 1   Customers        152 non-null    int64 
 2   Marketing_Spend  152 non-null    int64 
 3   Sales            152 non-null    int64 
dtypes: int64(3), object(1)
memory usage: 4.9+ KB


In [26]:
df.describe()

Customers  Marketing_Spend         Sales
count  152.000000       152.000000    152.000000
mean   112.927632      4922.368421  13750.657895
std     22.459412      1379.482933   3177.596382
min     70.000000      2350.000000   7600.000000
25%     94.000000      3687.500000  10675.000000
50%    115.000000      5200.000000  14875.000000
75%    130.000000      6100.000000  16100.000000
max    160.000000      7400.000000  19000.000000

In [27]:
df.isnull().sum()

Region             0
Customers          0
Marketing_Spend    0
Sales              0
dtype: int64

In [28]:
# Train/test split
X = df.drop("Sales", axis=1)
y = df["Sales"]

In [29]:
# Identify numerical + categorical features
num_features = ["Customers", "Marketing_Spend"]
cat_features = ["Region"]

In [30]:
# Build preprocessing pipeline
preprocessor = ColumnTransformer(
    transformers=[
        ("num", "passthrough", num_features),
        ("cat", OneHotEncoder(), cat_features)
    ]
)

In [31]:
# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [32]:
# Transform features
X_train_transformed = preprocessor.fit_transform(X_train)
X_test_transformed = preprocessor.transform(X_test)

In [33]:
# Fit model
model = RandomForestRegressor(random_state=42)
print(model.fit(X_train_transformed, y_train))

RandomForestRegressor(random_state=42)


In [34]:
# Predict
y_pred = model.predict(X_test_transformed)

In [35]:
# Calculate 
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print("RMSE:", rmse)

RMSE: 202.69831328769715


In [36]:
# Evaluate model
from sklearn.metrics import r2_score
print("R2 Score:", r2_score(y_test, y_pred))

R2 Score: 0.9948520509645178


### Results:
The regression model achieved an RMSE of 202.70, indicating low prediction error relative to the sales range. The model also produced an R² score of 0.9948, demonstrating excellent fit and strong predictive performance.

## Further Scope:

- Incorporate advanced feature engineering such as interaction terms, polynomial features, and domain-specific ratios to enhance model performance.
- Experiment with additional regression models (Random Forest, Gradient Boosting, XGBoost) and compare results using cross-validation.
- Automate the entire pipeline by integrating feature generation, model training, and evaluation into a reusable, production-ready workflow.